### PR TITLE
Rework of AttachCapabilitiesEvent

### PR DESCRIPTION
**Pull request by RealMangorage**
PR for reworking the AttachCapabillitiesEvent.

Currently in a prototype/testing stage. 
What is currently made is not final product. Changes may occur.

Documentation: Proposal (WIP)

We add a new class which handles all event posts for AttachCapabilitiesEvent<T> (Which now extends Event, and not GenericEvent<T>)

### **> CapabilitySystem.java**
```java
CapabilitySystem.addListener(Class<T> yourClassToListenFor, Consumer<AttachCapabilitiesEvent<T>> eventConsumer);

CapabilitySystem.addWrappedListener(Class<T> yourClassToListenFor, Class<W> classYourActuallyGonnaGet Consumer<AttachCapabilitiesEvent<W>> eventConsumer);

Examples:

CapabilitySystem.addListener(Player.class, e -> {}); // This will only be called and supplied with objects that are Player.class or extend Player.class.
 CapabilitySystem.addWrappedListener(ItemStack.class, BlockItem.class, ItemStack::getItem,  (event, item) -> {}); // This will only be called and supplied with objects who use BlockItem.class or extend BlockItem.class, and the objects it returns in this case will be ItemStack. Use .getItem() to get the item or use "item" to get it without needing to cast it.
```

### CODE REVIEW

**Comment by Jonathing** [comment]
Relevant conversation from Discord.

[](https://discord.com/channels/1129059589325852724/1129095235889270844/1154055409544462376)

File: patches/minecraft/net/minecraft/world/level/block/entity/BlockEntity.java.patch [hunk 1]
```diff
@@ -9,12 +9,12 @@
     private static final Logger f_58854_ = LogUtils.getLogger();
     private final BlockEntityType<?> f_58855_;
     @Nullable
-@@ -23,11 +_,14 @@
+@@ -23,11 +_,13 @@
     protected final BlockPos f_58858_;
     protected boolean f_58859_;
     private BlockState f_58856_;
+-
```
Comment (KnightMiner) [inline patches/minecraft/net/minecraft/world/level/block/entity/BlockEntity.java.patch]: This patch change looks wrong to me, looks like this PR added a trivial whitespace change to the patch

Comment (Jonathing) [reply patches/minecraft/net/minecraft/world/level/block/entity/BlockEntity.java.patch]: Agreed. Please remove it so we can keep the patch surface as low as possible.

**Review by Jonathing — CHANGES_REQUESTED** [review]
Effectively needs a rebase.

More info: https://discord.com/channels/1129059589325852724/1129059590156337154/1155564202853937283

File: src/main/java/net/minecraftforge/common/capabilities/CapabilitySystem.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import net.minecraft.world.item.Item;
+import net.minecraftforge.event.AttachCapabilitiesEvent;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+public class CapabilitySystem {
```
Comment (Jonathing) [inline src/main/java/net/minecraftforge/common/capabilities/CapabilitySystem.java]: Bracket on new line, please

File: src/main/java/net/minecraftforge/common/capabilities/CapabilitySystem.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    {
+        if (Item.class.isAssignableFrom(type)) throw new IllegalStateException("Unable to add Listener for Items. Use CapabilitySystem.addItemListener(ItemClass, Consumer)");
+        LISTENER_LIST.computeIfAbsent(type, (e) -> new ArrayList<>()).add((Consumer) eventConsumer);
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <T, W> void addWrappedListener(Class<T> type, Class<W> wrappedClass, BiConsumer<AttachCapabilitiesEvent<T>, W> eventConsumer)
+    {
+        if (!Item.class.isAssignableFrom(type)) throw new IllegalStateException("Unable to add Listener for Items. Use CapabilitySystem.addListener(Class, Consumer) for non Item related stuff");
+        LISTENER_LIST.computeIfAbsent(type, (e) -> new ArrayList<>()).add((cls) -> {
```
Comment (Jonathing) [inline src/main/java/net/minecraftforge/common/capabilities/CapabilitySystem.java]: I don't remember if lambda brackets also go on new line. @LexManos?

File: src/main/java/net/minecraftforge/common/extensions/IForgeEntity.java [hunk 4]
```diff
@@ -448,4 +448,4 @@ default boolean shouldUpdateFluidWhileBoating(FluidState state, Boat boat)
     {
         return boat.shouldUpdateFluidWhileRiding(state, self());
     }
-}
```
Comment (Jonathing) [inline src/main/java/net/minecraftforge/common/extensions/IForgeEntity.java]: Unnecessary whitespace edit

File: src/main/java/net/minecraftforge/common/extensions/IForgeLevelChunk.java [hunk 5]
```diff
@@ -10,4 +10,4 @@
 public interface IForgeLevelChunk extends ICapabilityProvider
 {
 
-}
```
Comment (Jonathing) [inline src/main/java/net/minecraftforge/common/extensions/IForgeLevelChunk.java]: Unnecessary whitespace edit

File: src/main/java/net/minecraftforge/common/ForgeHooks.java [hunk 6]
```diff
@@ -1250,5 +1250,4 @@ public static void handleClientConfigurationComplete(Connection connection) {
         } else
             LOGGER.info("Connected to a modded server.");
     }
-
 }
```
Comment (Jonathing) [inline src/main/java/net/minecraftforge/common/ForgeHooks.java]: Unnecessary whitespace edit

File: src/main/java/net/minecraftforge/fluids/capability/wrappers/FluidBucketWrapper.java [hunk 7]
```diff
@@ -174,4 +174,4 @@ public <T> LazyOptional<T> getCapability(@NotNull Capability<T> capability, @Nul
     {
         return ForgeCapabilities.FLUID_HANDLER_ITEM.orEmpty(capability, holder);
     }
-}
```
Comment (Jonathing) [inline src/main/java/net/minecraftforge/fluids/capability/wrappers/FluidBucketWrapper.java]: Unnecessary whitespace edit

File: src/main/java/net/minecraftforge/event/AttachCapabilitiesEvent.java [hunk 8]
```diff
@@ -53,8 +59,7 @@ public T getObject()
      */
     public void addCapability(ResourceLocation key, ICapabilityProvider cap)
     {
-        if (caps.containsKey(key))
-            throw new IllegalStateException("Duplicate Capability Key: " + key  + " " + cap);
```
Comment (Jonathing) [inline src/main/java/net/minecraftforge/event/AttachCapabilitiesEvent.java]: Unnecessary whitespace edit (unless this is an intentional part of the refactor)

File: src/test/java/net/minecraftforge/debug/CapabilitiesTest.java
Comment (Jonathing) [inline src/test/java/net/minecraftforge/debug/CapabilitiesTest.java:1]: Similar formatting issues to `CapabilitySystem.java`. Inconsistent usage of new-line brackets.

File: src/test/resources/META-INF/mods.toml [hunk 9]
```diff
@@ -4,10 +4,13 @@ license="LGPL v2.1"
...
 [[mods]]
 modId="empty"
+[[mods]]
+modId="captest"
 
 [[mods]]
 modId="creative_mode_tab_test"
-
+[[mods]]
+modId="captest"
```
Comment (Jonathing) [inline src/test/resources/META-INF/mods.toml]: I think the test mod only needs to be declared in this file once.

File: src/test_old/java/net/minecraftforge/debug/CapabilitiesTest.java [hunk 10]
```diff
@@ -40,26 +42,44 @@ public class CapabilitiesTest
 {
     public static final String MODID = "capabilities_test";
 
-    private static final boolean ENABLED = false;
+    private static final boolean ENABLED = true;
```
Comment (Jonathing) [inline src/test_old/java/net/minecraftforge/debug/CapabilitiesTest.java:45]: Don't forget to disable this after the PR is out of draft state

**Review by Jonathing — CHANGES_REQUESTED** [review]

File: src/main/java/net/minecraftforge/common/capabilities/CapabilitySystem.java
Comment (Jonathing) [inline src/main/java/net/minecraftforge/common/capabilities/CapabilitySystem.java:1]: This class needs javadocs, especially since it looks like modders are going to be interfacing with this.

File: src/test/java/net/minecraftforge/debug/CapabilitiesTest.java [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+@Mod(CapabilitiesTest.MOD_ID)
+public class CapabilitiesTest
+{
+    public static final String MOD_ID = "captest";
+    private static final ConcurrentHashMap<Class<?>, AtomicInteger> TRACK = new ConcurrentHashMap<>();
+    private static final boolean ENABLED = true;
```
Comment (Jonathing) [inline src/test/java/net/minecraftforge/debug/CapabilitiesTest.java:45]: Remember to set this to false before marking the PR as ready!

**Review by Jonathing — CHANGES_REQUESTED** [review]
Good shit so far. It works *a lot* better than the horse shit we had before, especially with the explicit class listening. Obviously it needs a lot more testing, but it's headed in the right direction. Like I said in VC, I am happy to help with remedial shit like javadocs and the primer.

**Comment by Jonathing** [comment]
Superseded by #10118.
